Adds documentation for antialiasing with impeller (#13370)
flutter/website: sites/docs/src/content/perf/antialiasing.md

---
title: Impeller anti-aliasing
description: How does Impeller perform anti-aliasing?
---

> Aliasing is the visual artifacts that result from drawing geometry to a grid of

Reviewer: Optional nit:

If I weren't familiar with the problem, I'm not sure I would understand what issue is being described here. 
What visual artifacts?
An image or two could add a lot of value here. Maybe we could pull from some of the issues we've been solving for impeller macOS. But that will only help once folks get to this page.

To improve the chances that a search for the issue would land devs on this page, it may also help to add a few other descriptions of aliasing. Artifacts is a good term, but maybe adding a few synonyms could help, such as jagged/rough/hard edges ("jaggies"), stair-stepping, spatial sampling errors

Author: Good call, done.

---
title: Impeller anti-aliasing
description: How does Impeller perform anti-aliasing?
---

Aliasing is the visual artifacts that result from drawing geometry to a grid of

> pixels (rasterization). Impeller employs a couple of techniques to smooth out

Reviewer 2: ```suggestion
pixels (rasterization). Impeller employs a couple techniques to smooth out
```

Author: done

the mapping to raster graphics (anti-aliasing).

## Techniques

### Multisample anti-aliasing (MSAA)

[MSAA][] is a global anti-aliasing technique that operates on the whole contents

> of the screen. It is an optimization over rendering the whole screen at a larger

Reviewer 2: ```suggestion
It's more optimized than rendering the whole screen at a larger
```

Author: The meaning I'm trying to convey is that this is a technique that is an optimization on supersampling.  I'm spelling out supersampling though to try to make it more accessible.  The suggestion loses that relationship.

of the screen. It is an optimization over rendering the whole screen at a larger
scale and shrinking it down ([SSAA][]). Instead of doing the fragment operation
for each fragment in a region, if it is determined they have the same coverage,
only one fragment operation is calculated. This limits smoothing to edges.
Mobile phone GPUs have special hardware to optimize this process (
[Tiled rendering][]). It comes in varying degrees of how many samples to

> consider.

Reviewer 3: Not sure how you are defining "fragment operation". If it is "execution of a fragment program" then I don't think this is correct.

For any given pixel I believe it only performs one fragment shader operation on the center of a pixel for any pixel, whether it has all the same MSAA coverage or not. The only thing that MSAA adds is which of the sub-samples it propagates that single result to.

The text above implies that it only runs the fragment shader once for wholly enclosed interior pixels - and??? runs it multiple times for edge pixles? (I don't think it works that way).

Author: Clarified, thanks.

> On desktop and mobile 4x MSAA is used for all rendering calls.

Reviewer 3: Perhaps mention the consequences of the MSAA #, such as "the quantization of the coverage for edge pixels is limited to the number of MSAA samples" (so edge pixels are always only one of 0%, 25%, 50%, 75%, or 100% covered with MSAA 4).

Author: I played around with this, I don't think we need to go into this level of detail.

### Signed distance fields ([SDFs][])

Typically, hardware accelerated computer graphics is done by defining a series

> of points and edges (a [mesh][]) and [shaders][]. SDF rendering instead renders

Reviewer 2: ```suggestion
Typically, hardware accelerated computer graphics define a series
of points and edges (as a [mesh][]) and [shaders][]). Instead, SDF renders
```

Author: Done.  I did keep "a mesh" though since it is defining the concept.  It's not an example of how those things could be a collection.

> edges (a [mesh][]) and [shaders][]. Instead, SDF renders shapes in the fragment

Reviewer 3: "... and the color sample for pixels is either 0% or 100% the result of the fragment shader (color source?) for each pixel depending on whether it falls in the mesh or not"...?

Author: (this is a companion to the msaa 4x comment, see above)

> shape of the object being drawn. Since the shape is defined in the fragment

Reviewer 2: ```suggestion
shapes in the fragment shader program as signed distance fields.
Since the shape is defined in the fragment
```

Author: done

fragment shader the edges can be smoothed at the fragment level instead of

> relying on the rasterization of a mesh.

Reviewer 3: "smoothed ... by computing the coverage of the shape over each individual pixel at a high resolution (or precision?) and used to proportionally apply the color sample to that pixel"...?

Author: I don't think this really make it more clear.

> of relying on the rasterization of a mesh.

Reviewer 2: ```suggestion
shader, the edges can be smoothed at the fragment level instead
of relying on the rasterization of a mesh.
```

Author: done

> On desktop, rendering with SDFs is enabled. On mobile platforms, SDFs are an

Reviewer 3: "... enabled by default"?

Author: done

> option whose default value is false.

Reviewer 2: ```suggestion
On desktop, rendering with SDFs is enabled. On mobile platforms,
SDF is an option that defaults to false.
```

Author: done

This technique is prioritized on desktop because SDF rendering puts more demand
on the GPU and Flutter supports older mobile phones. Also, the physical pixel

> sizes on desktop computers are typically bigger than those of mobile phones. So

Reviewer 3: Desktop computers don't have pixels. ;)

"desktop displays"?

Perhaps provide some example statistics?

Reviewer 3: Also, bigger in terms of field of view. Mobile phones are viewed closer than desktop displays, but their much higher pixel count still wins out in terms of pixel angular viewing angle thingy bopper...

Author: Ack, I don't know if we need to go into that detail.

any imperfection will be more evident there.

### Examples

| No AA | MSAA 4x | MSAA 4x + SDF |
| --- | --- | --- |
| ![No AA](/assets/images/docs/perf/noaa.png) | ![MSAA 4x](/assets/images/docs/perf/msaa4.png) | ![MSAA 4x + SDF](/assets/images/docs/perf/msaa4sdf.png) |

## Working with anti-aliasing

### SDFs with the FragmentShader API

> Standard primitive shapes in Flutter will be drawn automatically with SDFs. If

Reviewer 2: ```suggestion
Standard primitive shapes in Flutter are drawn automatically with SDFs. If
```

Author: done

> | ![No AA](/assets/images/docs/perf/noaa.png) | ![MSAA 4x](/assets/images/docs/perf/msaa4.png) | ![MSAA 4x + SDF](/assets/images/docs/perf/msaa4sdf.png) |

Reviewer 3: The images are subtle. One other option is to show a very slightly rotated line where the stepping is far more abrupt with MSAA and it is easy to see more steps with SDF.

Author: ack, if you have some can you file a folow up pr?

> Standard primitive shapes in Flutter are drawn automatically with SDFs. If a

Reviewer 3: automatically -> by default unless the developer changes the "?antialiased?" setting in the Paint object...?

Author: updated wording, the doc doesn't mention the flag, I'd want to play around with it more before mentioning it to make sure we cover all cases

a Flutter developer wants to define their own custom graphics with SDFs they can
do so with the [FragmentShader API][]. Using the [drawPath()][] is sufficient
for most use cases without resorting to high quality SDF rendering. Not all

> drawn paths are guaranteed to result in SDF rendering though.

Reviewer 2: Do we want to say more about this? Why might it be ignored even if disabled?

Author: It's the inverse of that, you have SDFs enabled but you aren't guaranteed to get them always.  I'm sort of covering for the current limitations in the engine.  It's an ongoing research problem.  There will always be some frontier where it becomes impractical because of hardware limitations.  I don't know if I want to wade into that.
